File: 11004/main.py
import sys

N, K = (int(term) for term in sys.stdin.readline().split())
A = sys.stdin.readline().split()
for i in range(N):
    A[i] = int(A[i])

K -= 1

def partition(begin, end):
    lo, pi = begin, end-1
    pivot = A[pi]
    for i in range(begin, end-1):
        if A[i] <= pivot:
            A[lo], A[i] = A[i], A[lo]
            lo += 1
    A[lo], A[pi] = A[pi], A[lo]
    return lo


def f(begin, end):
    lo = partition(begin, end)

    if lo < K:
        return f(lo+1, end)

    if lo > K:
        return f(begin, lo)

    return A[K]

print(f(0, N))

File: 11004/main.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"time"
)

func main() {
	rand.Seed(time.Now().UTC().UnixNano())

	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)
	nextInt := func() int {
		_ = scanner.Scan()
		n, _ := strconv.Atoi(scanner.Text())
		return n
	}

	N := nextInt()
	K := nextInt() - 1

	A := make([]int, N)
	for i := 0; i < N; i++ {
		A[i] = nextInt()
	}

	begin, end := 0, N
	for begin < end {
		i := begin
		pivot := A[begin]
		for j := begin + 1; j < end; j++ {
			if A[j] <= pivot {
				i++
				A[i], A[j] = A[j], A[i]
			}
		}

		A[begin], A[i] = A[i], A[begin]

		if i < K {
			begin = i + 1
			continue
		}

		if i > K {
			end = i
			continue
		}

		break
	}

	fmt.Println(A[K])
}
